java/java-impl-inspections/src/com/intellij/codeInspection/RedundantExplicitCloseInspection.java

```diff
@@ -29,40 +32,58 @@ public void visitTryStatement(@NotNull PsiTryStatement statement) {
+        for (PsiStatement last : terminetingStatements) {
+          PsiExpressionStatement expressionStatement = tryCast(last, PsiExpressionStatement.class);
+          if(expressionStatement == null) return;
+          PsiMethodCallExpression call = tryCast(expressionStatement.getExpression(), PsiMethodCallExpression.class);
+          if (!CLOSE.test(call)) return;
+          PsiExpression qualifier = call.getMethodExpression().getQualifierExpression();
+          PsiReferenceExpression reference = tryCast(PsiUtil.skipParenthesizedExprDown(qualifier), PsiReferenceExpression.class);
+          if(reference == null) return;
+          PsiVariable variable = tryCast(reference.resolve(), PsiVariable.class);
+          if(variable == null) return;
+          boolean isReferenceToResourceVariable = StreamEx.of(resourceList.iterator()).anyMatch(
+            element -> variable == element ||
+                       element instanceof PsiResourceExpression &&
+                       EquivalenceChecker.getCanonicalPsiEquivalence()
+                         .expressionsAreEquivalent(reference, ((PsiResourceExpression)element).getExpression()));
+          if(!isReferenceToResourceVariable) return;
+          holder.registerProblem(last, JavaBundle.message("inspection.redundant.explicit.close"), new DeleteElementFix(last, CommonQuickFixBundle.message("fix.remove.redundant", "close()")));
+        }
       }
     };
   }
 
-  private static class DeleteRedundantCloseFix implements LocalQuickFix {
-    @Nls
-    @NotNull
-    @Override
-    public String getFamilyName() {
-      return CommonQuickFixBundle.message("fix.remove.redundant", "close()");
+  @NotNull
+  private static List<PsiStatement> getTerminatingStatements(@Nullable PsiStatement last) {
+    if (last == null) return Collections.emptyList();
+    List<PsiStatement> terminetingStatements = new ArrayList<>();
+    PsiIfStatement ifStatement = tryCast(last, PsiIfStatement.class);
+    if (ifStatement != null) {
+      PsiStatement[] thenStatements = getStatements(ifStatement.getThenBranch());
+      terminetingStatements.addAll(getTerminatingStatements(ArrayUtil.getLastElement(thenStatements)));
+      PsiStatement elseBranch = ifStatement.getElseBranch();
+      if (elseBranch != null) {
+        PsiStatement[] elseStatements = getStatements(elseBranch);
+        terminetingStatements.addAll(getTerminatingStatements(ArrayUtil.getLastElement(elseStatements)));
+      }
+    } else {
+      terminetingStatements.add(last);
     }
+    return terminetingStatements;
+  }
 
-    @Override
-    public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
-      PsiElement element = descriptor.getStartElement();
-      new CommentTracker().deleteAndRestoreComments(element);
+  @NotNull
+  private static PsiStatement[] getStatements(@Nullable PsiStatement branch) {
```

Comment:
It looks like, com.siyeh.ig.psiutils.ControlFlowUtils#unwrapBlock does the same.

Reply:
Applied